backend/headless/output.c

```diff
@@ -0,0 +1,146 @@
+#include <stdlib.h>
+#include <EGL/egl.h>
+#include <EGL/eglext.h>
+#include <GLES2/gl2.h>
+#include <wlr/interfaces/wlr_output.h>
+#include <wlr/util/log.h>
+#include "backend/headless.h"
+
+static EGLSurface egl_create_surface(struct wlr_egl *egl, unsigned int width,
+		unsigned int height) {
+	EGLint attribs[] = {EGL_WIDTH, width, EGL_HEIGHT, height, EGL_NONE};
+
+	EGLSurface surf = eglCreatePbufferSurface(egl->display, egl->config, attribs);
+	if (surf == EGL_NO_SURFACE) {
+		wlr_log(L_ERROR, "Failed to create EGL surface: %s", egl_error());
+		return EGL_NO_SURFACE;
+	}
+	return surf;
+}
+
+static bool output_set_custom_mode(struct wlr_output *wlr_output, int32_t width,
+		int32_t height, int32_t refresh) {
+	struct wlr_headless_output *output =
+		(struct wlr_headless_output *)wlr_output;
+	struct wlr_headless_backend *backend = output->backend;
+
+	if (output->egl_surface) {
+		eglDestroySurface(backend->egl.display, output->egl_surface);
+	}
+
+	output->egl_surface = egl_create_surface(&backend->egl, width, height);
+	if (output->egl_surface == EGL_NO_SURFACE) {
+		wlr_log(L_ERROR, "Failed to recreate EGL surface");
+		wlr_output_destroy(wlr_output);
+		return false;
+	}
+
+	output->frame_delay = 1000000 / refresh;
+
+	wlr_output_update_custom_mode(&output->wlr_output, width, height, refresh);
+	return true;
+}
+
+static void output_transform(struct wlr_output *wlr_output,
+		enum wl_output_transform transform) {
+	struct wlr_headless_output *output =
+		(struct wlr_headless_output *)wlr_output;
+	output->wlr_output.transform = transform;
+}
+
+static void output_make_current(struct wlr_output *wlr_output) {
+	struct wlr_headless_output *output =
+		(struct wlr_headless_output *)wlr_output;
+	if (!eglMakeCurrent(output->backend->egl.display,
+		output->egl_surface, output->egl_surface,
+		output->backend->egl.context)) {
+		wlr_log(L_ERROR, "eglMakeCurrent failed: %s", egl_error());
+	}
+}
+
+static void output_swap_buffers(struct wlr_output *wlr_output) {
+	// No-op
```

Comment:
Are you sure you don't need to call eglSwapBuffers here?

Reply:
From https://www.khronos.org/registry/EGL/sdk/docs/man/html/eglSwapBuffers.xhtml

>If surface is a pixel buffer or a pixmap, eglSwapBuffers has no effect, and no error is generated.